Azure Blob storage support in buildcache

garvct opened the pull request:
Added support for Azure Blob storage to be used as a buildcache mirror.

Support the following syntax.

spack mirror add azure_buildcache azure://<STORAGE_ACCOUNT_NAME>.blob.core.windows.net/<CONTAINER_NAME>

Then
spack buildcache create --rebuild-index -k my_gpg -m azure_buildcache <SPEC>
spack buildcache install <SPEC>

**Prerequisites**
An Azure storage connection string needs to be defined via the AZURE_STORAGE_CONNECTION_STRING environmental variable.
The Azure blob storage python SDK needs to be imported.
https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-python

garvct committed:
Added azure_blob.py

garvct committed:
Added Azure blob storage integration to web.py

garvct committed:
Added Azure blob storage integration to fetch_strategy.py

garvct committed:
Corrected formatting errors

garvct committed:
Removed pathlib

garvct committed:
formatted azure_blob.py

garvct committed:
Correct format error

garvct committed:
Removed sys.exit()

garvct committed:
fixed syntax error

garvct committed:
Corrected generate_account_sas syntax

garvct committed:
Corrected file formatting

garvct committed:
Corrected file formatting

garvct committed:
Changed format of web.py

garvct committed:
Changed code format

garvct committed:
Changed code format

garvct committed:
Changed code format

garvct committed:
Changed code format

garvct committed:
Changed code format

garvct committed:
Changed code format

garvct committed:
Changed code format

garvct committed:
Changed code format

garvct committed:
Changed code format

garvct committed:
Changed code format

garvct committed:
Added azure_delete_blob and modified azure_blob_list

garvct commented:
@tgamblin @adamjstewart @gartung please review when you get a chance. Thank you.

garvct committed:
Moved azure code above check if url exists

tgamblin commented:
Thanks for this contribution!  I’ll try to get to this, or possibly ask @opadron to look at it. He may be mostly out til January, though.

xpillons commented:
@opadron when do you think this merge will be done ?

alalazo commented:
@opadron @garvct Is this still something we want to get merged (I assume so, but want to double check)?

garvct commented:
Yes, I would like to get this merged. What can I do to help the process forward?

alalazo commented:
@garvct Can you please rebase the PR on the latest `develop`? I'll try to take a first look at it before pinging @opadron for a final review. Thanks!

garvct committed:
moved location of azure_blob import

garvct committed:
reordered the imports

garvct committed:
reorder imports in azure_url_sas.

garvct committed:
added comma imports in azure_url_sas.

garvct committed:
Updated license header.

garvct commented:
@alalazo @opadron I have rebased this branch with the latest develop branch, please review at your convenience.

dependabot[bot] committed:
build(deps): bump docker/build-push-action from 2.8.0 to 2.9.0

Bumps [docker/build-push-action](https://github.com/docker/build-push-action) from 2.8.0 to 2.9.0.
- [Release notes](https://github.com/docker/build-push-action/releases)
- [Commits](https://github.com/docker/build-push-action/compare/1814d3dfb36d6f84174e61f4a4b05bd84089a4b9...7f9d37fa544684fb73bfe4835ed7214c255ce02b)

---
updated-dependencies:
- dependency-name: docker/build-push-action
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

alalazo commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -318,6 +318,11 @@ def fetch(self):
         url = None
         errors = []
         for url in self.candidate_urls:
+            if url[0:5] == 'azure':
```
If `url` is a string it's probably better to:
```suggestion
            if url.startswith('azure'):
```

garvct replied:
I changed the syntax.

alalazo commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -318,6 +318,11 @@ def fetch(self):
         errors = []
         for url in self.candidate_urls:
+            if url[0:5] == 'azure':
+                import spack.util.azure_blob as azure_blob_util
```
This import might become an issue here, since an import is an implicit assignment of a variable. Here a simple example with the same structure:
```python
Python 3.8.10 (default, Nov 26 2021, 20:14:08) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> def foo():
...   print('Before')
...   dir(os)
...   import os.path
>>> foo()
Before
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in foo
UnboundLocalError: local variable 'os' referenced before assignment
```

garvct replied:
I think the import spack.util.azure_blob as azure_blob_util will work, similar syntx is used throughput the fetch_strategy.py file.
(e.g import spack.util.s3 as s3_util)

alalazo commented on lib/spack/spack/util/azure_blob.py:
```diff
@@ -0,0 +1,135 @@
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+"""This file contains the definition of the Azure Blob storage Class used to
+   integrate Azure Blob storage with spack buildcache.
+"""
```
Would it make sense to call this file just `azure.py`? Asking since I don't know much the domain so I might be missing details why it's better not to.

garvct replied:
azure may be too general? azure_blob is specific to azure blob storage only.

alalazo commented on lib/spack/spack/util/azure_blob.py:
```diff
@@ -0,0 +1,135 @@
+        from azure.storage.blob import BlobServiceClient
+        self.url = url
+        (self.container_name, self.blob_path) = self.get_container_blob_path()
+        if url.scheme != 'azure':
+            raise ValueError(
+                'Can not create Azure blob connection from URL with scheme: %s'
+                % (url.scheme))
```
Can we move input checking to the beginning of the method? I usually find it better, since when reading code one can assume that after the checks inputs are valid.

alalazo commented on lib/spack/spack/util/azure_blob.py:
```diff
@@ -0,0 +1,135 @@
+            raise ValueError(
+                'Can not create Azure blob connection from URL with scheme: %s'
+                % (url.scheme))
+        if "AZURE_STORAGE_CONNECTION_STRING" in os.environ:
```
Can we pass the connection string as an argument to the constructor? That will probably make the class easier to deal with in tests. We can then have a small factory function or something similar that reads the connection string from the environment and injects it into a newly created object.

alalazo commented on lib/spack/spack/util/azure_blob.py:
```diff
@@ -0,0 +1,135 @@
+ required if you want to use Azure Blob storage as an \
+ spack buildcache")
+
+    def get_container_blob_path(self):
+        blob_path = self.url.path
+        path_list = blob_path.split('/')
+        container_name = path_list[1]
+        blob_path = os.path.join(*path_list[2:])
+        tty.debug("container_name = %s, blob_path = %s"
+                  % (container_name, blob_path))
+        return (container_name, blob_path)
```
Editing in place, so beware of typos. Also these are just suggestions so feel free to take what seems sensible to you:
```suggestion
    def container_and_blob(self):
        """Return the container name and the blob path from the full url."""
        path_components = self.url.path.split('/')
        container_name = path_components[1]
        blob_path = os.path.join(*path_components[2:])
        
        msg = "container_name = {0}, blob_path = {1}"
        tty.debug(msg.format(container_name, blob_path))
        
        return container_name, blob_path
```

alalazo commented on lib/spack/spack/util/azure_blob.py:
```diff
@@ -0,0 +1,135 @@
+        blob_path = self.url.path
+        path_list = blob_path.split('/')
+        container_name = path_list[1]
+        blob_path = os.path.join(*path_list[2:])
```
Is it really `os.path.join`? Meaning, do we have to use different path separators depending on the OS or can we assume it's always `/` like for URLs?

alalazo commented on lib/spack/spack/util/azure_blob.py:
```diff
@@ -0,0 +1,135 @@
+        return (container_name, blob_path)
+
+    def is_https(self):
+        if 'https' in self.connect_str:
+            return True
+        else:
+            return False
```
Would this one work:
```suggestion
        return self.connect_str.startswith('https')
```
?

garvct replied:
'https' is not at the beginning of the connection string. (startswith will not work in this case)

alalazo commented on lib/spack/spack/util/azure_blob.py:
```diff
@@ -0,0 +1,135 @@
+            return False
+
+    def azure_container_exists(self):
+        try:
+            ttt = (self.blob_service_client.
+                   get_container_client(self.container_name).list_blobs())
+            for blob in ttt:
+                pass
+        except Exception:
+            return False
+        return True
```
Can you explain somewhere what this method does? In particular why iterating over an object without performing any operation?

garvct replied:
I added a comment in the code.

alalazo commented on lib/spack/spack/util/azure_blob.py:
```diff
@@ -0,0 +1,135 @@
+            return False
+        return True
+
+    def azure_blob_exists(self):
```
It might be good to have a brief docstring explaining why a call to `get_properties` ensures that the blob exists.

garvct replied:
I added a comment.

alalazo reviewed (changes requested):
I did a first pass over the PR and left a few comments. Let me know if you have questions. In general I think it would be good to have brief docstrings for methods, if they are part of the public API.

dependabot[bot] committed:
build(deps): bump actions/setup-python from 2.3.1 to 2.3.2

Bumps [actions/setup-python](https://github.com/actions/setup-python) from 2.3.1 to 2.3.2.
- [Release notes](https://github.com/actions/setup-python/releases)
- [Commits](https://github.com/actions/setup-python/compare/f38219332975fe8f9c04cca981d674bf22aea1d3...7f80679172b057fc5e90d70d197929d454754a5a)

---
updated-dependencies:
- dependency-name: actions/setup-python
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

garvct committed:
Merge pull request #1 from garvct/dependabot/github_actions/docker/build-push-action-2.9.0

build(deps): bump docker/build-push-action from 2.8.0 to 2.9.0

garvct committed:
Merge pull request #2 from garvct/dependabot/github_actions/actions/setup-python-2.3.2

build(deps): bump actions/setup-python from 2.3.1 to 2.3.2

dependabot[bot] committed:
build(deps): bump docker/login-action from 1.12.0 to 1.13.0

Bumps [docker/login-action](https://github.com/docker/login-action) from 1.12.0 to 1.13.0.
- [Release notes](https://github.com/docker/login-action/releases)
- [Commits](https://github.com/docker/login-action/compare/42d299face0c5c43a0487c477f595ac9cf22f1a7...6af3c118c8376c675363897acf1757f7a9be6583)

---
updated-dependencies:
- dependency-name: docker/login-action
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

garvct committed:
Merge branch 'spack:develop' into develop

garvct committed:
Merge pull request #3 from garvct/dependabot/github_actions/docker/login-action-1.13.0

build(deps): bump docker/login-action from 1.12.0 to 1.13.0

garvct committed:
Used str.startswith syntax

garvct committed:
Moved arg error check

garvct committed:
Updated function get_container_blob_path

garvct committed:
Added comment about azure_container_exists function.

garvct committed:
Added comment about azure_blob_exists function.

ltalirz commented:
hi @gartung , I see you approved this PR in June 2022; is there anything stopping the PR from being merged (besides rebasing on top of `develop`)?

we would also be interested in this feature

gartung commented:
Ask @tgamblin

marconetto commented:
Hi folks, I'm interested in this feature, and wondering if there are blockers to move this forward.

ltalirz commented:
Hi @marconetto , that sounds great!

No blockers, just time.

The spack code changed in a couple of places since this original PR, so I wasn't able to finish the implementation in the time I had available.

I currently don't find my trials; not sure I still have them
If you would like any input from my side, let me know
